crates/phactory/src/system/gk.rs

```diff
@@ -911,7 +911,25 @@ where
                 challenge_block,
                 challenge_time,
                 iterations,
-            } => {
+            } |
+            MiningReportEvent::HeartbeatV2 {
+                session_id,
+                challenge_block,
+                challenge_time,
+                iterations,
+                ..
+            }
+             => {
+                let n_contracts = if let MiningReportEvent::HeartbeatV2 {
+                    n_contracts,
+                    ..
+                } = event {
+                    n_contracts
+                } else {
+                    0
+                };
+                // TODO.kevin: Deal with workers running contracts.
```

Comment:
@h4x3rotab How do we deal with workers running contracts?

Reply:
Haven't thought deeply. Generally we want the miner to keep get the same reward as mining normally. We use the iteration count to calculate `p_instant`, which affects the actual reward. When the miner is hosting the contract, we can turn its mining state from `MiningIdle` to `MiningActive`, and just give it the reward as 100% `p_init`. When the miner returns back to `MiningIdle`, we can refresh the iterator and make it back to normal.

Reply:
Btw, I saw `n_contracts` and `n_clusters`. I think we should treat `n_clusters > 0` as "active"

Reply:
Treating n_clusters > 0 as "active" and Using p_init to calculate the reword. Done